staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -629,9 +629,16 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		c.watcherIdx++
 	}()
 
-	// FIXME: if empty watchcache, we won't have events, some may be in the queue so delivered,
-	//   but we're not passing resourceVersion correctly...
-	go watcher.processInterval(ctx, cacheInterval, watchRV)
+	// Some events might have already been bufferred and not yet processed
+	// by the cacher. By passing the resource version after the initEvents,
+	// we ensure that hose will not be delivered to the watcher for the
+	// second time, causing it going back in time.
+	// At the same time it's possible that watchRV is still more to the
+	// future then afterInitEventsRV, so we need to take a max of those.
+	if watchRV > afterInitEventsRV {
+		afterInitEventsRV = watchRV
```

Comment:
it would be clearer to assign to a separate variable rather than modifying `afterInitEventsRV`